impersonate/middleware.py

```diff
@@ -1,26 +1,55 @@
+from django.utils.functional import empty, SimpleLazyObject
 from .helpers import User, check_allow_for_uri, check_allow_for_user
 
 
-class ImpersonateMiddleware(object):
+def apply_impersonate(request):
+    request.user.is_impersonate = False
+    request.__dict__['impersonator'] = None
+
+    if request.user.is_authenticated() and \
+       '_impersonate' in request.session:
+        new_user_id = request.session['_impersonate']
+        if isinstance(new_user_id, User):
+            # Edge case for issue 15
+            new_user_id = new_user_id.id
+
+        try:
+            new_user = User.objects.get(id=new_user_id)
+        except User.DoesNotExist:
+            return
+
+        if check_allow_for_user(request, new_user) and \
+           check_allow_for_uri(request.path):
+            request.__dict__['impersonator'] = request.user
+            request.user = new_user
+            request.user.is_impersonate = True
+
+
+def impersonator(request):
+    # trigger apply_impersonate
+    request.user.is_authenticated()
+
+    return request.impersonator
 
+
+class ImpersonateMiddleware(object):
```

Comment:
This code used to be pretty straight forward. I hope this optimization is worth it. I also hope django-impersonate accepts this into their code base.